lua/aerial/backends/util.lua

```diff
@@ -55,4 +55,59 @@ M.remove_change_watcher = function(bufnr, backend_name)
   })
 end
 
+-- Taken directly out of nvim-treesitter with minor adjustments
+if (vim.treesitter.language == nil) or (vim.treesitter.language.get_lang == nil) then
+  ---@param bufnr nil|integer
+  M.get_buf_lang = function(bufnr)
+    local ok, parsers = pcall(require, "nvim-treesitter.parsers")
+    if not ok then
+      return false, "nvim-treesitter not found"
+    end
+    return parsers.get_buf_lang(bufnr)
+  end
+else
+  ---@param bufnr nil|integer
+  M.get_buf_lang = function(bufnr)
+    bufnr = bufnr or vim.api.nvim_get_current_buf()
+    local ft = vim.api.nvim_buf_get_option(bufnr, "ft")
+
+    local result = vim.treesitter.language.get_lang(ft)
+    if result then
+      return result
+    else
+      ft = vim.split(ft, ".", { plain = true })[1]
+      return vim.treesitter.language.get_lang(ft) or ft
+    end
+  end
+end
+
+if (vim.treesitter.language == nil) or (vim.treesitter.language.get_filetypes == nil) then
+  ---@param lang string
+  M.has_parser = function(lang)
+    return require("nvim-treesitter.parsers").has_parser(lang)
+  end
+else
+  ---@param lang string
+  M.has_parser = function(lang)
+    if next(vim.treesitter.language.get_filetypes(lang)) ~= nil then
```

Comment:
In my test `get_filetypes()` always returns a non-empty table, even if the parser is not installed. I doubt if this function can be used.

Reply:
Oh, you're absolutely right. I'd switched to using `get_string_parser` and recovering from error there. Even nvim-treesitter uses private APIs for now, so this might be a good enough solution until a better API becomes available.

Alternatively, `has_parser` and `get_buf_lang` can be rid of.
Starting with nvim 0.9, it's possible to simply grab a parser by buffer number, letting neovim figure it out.